<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Copy number variation analysis with Bioconductor • enrichOmics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Copy number variation analysis with Bioconductor">
<meta property="og:description" content="CNVWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">enrichOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Geistlinger_CNVWorkshop.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/CNVWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Copy number variation analysis with Bioconductor</h1>
                        <h4 class="author">Ludwig Geistlinger</h4>
            <address class="author_afil">
      CUNY School of Public Health, New York, NY<br><h4 class="author">Levi Waldron</h4>
            <address class="author_afil">
      CUNY School of Public Health, New York, NY<br><h4 class="date">June 08, 2020</h4>
      
      
      <div class="hidden name"><code>Geistlinger_CNVWorkshop.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div id="copy-number-variation-analysis-with-bioconductor" class="section level1">
<h1 class="hasAnchor">
<a href="#copy-number-variation-analysis-with-bioconductor" class="anchor"></a>Copy number variation analysis with Bioconductor</h1>
<div id="workshop-information" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-information" class="anchor"></a>Workshop information</h2>
<div id="instructors-names-and-contact-information" class="section level3">
<h3 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h3>
<p>Ludwig Geistlinger, Marcel Ramos, Sehyun Oh, and Levi Waldron</p>
<p>CUNY School of Public Health 55 W 125th St, New York, NY 10027</p>
<p><a href="mailto:Ludwig.Geistlinger@sph.cuny.edu" class="email">Ludwig.Geistlinger@sph.cuny.edu</a> <a href="mailto:Marcel.Ramos@sph.cuny.edu" class="email">Marcel.Ramos@sph.cuny.edu</a> <a href="mailto:Sehyun.Oh@sph.cuny.edu" class="email">Sehyun.Oh@sph.cuny.edu</a> <a href="mailto:Levi.Waldron@sph.cuny.edu" class="email">Levi.Waldron@sph.cuny.edu</a></p>
</div>
<div id="workshop-description" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h3>
<p>This workshop gives an overview of Bioconductor solutions for the analysis of copy number variation (CNV) data. The workshop introduces Bioconductor core data structures for efficient representation, access, and manipulation of CNV data, and how to use these containers for structured downstream analysis of CNVs and integration with gene expression and quantitative phenotypes. Participants will be provided with code and hands-on practice for a comprehensive set of typical analysis steps including exploratory analysis, summarizing individual CNV calls across a population, overlap analysis with functional genomic regions and regulatory elements, expression quantitative trait loci (eQTL) analysis, and genome-wide association analysis (GWAS) with quantitative phenotypes. As an advanced application example, the workshop also introduces allele-specific absolute copy number analysis and how it is incorporated in genomic cancer analysis for the estimation of tumor characteristics such as tumor purity and ploidy.</p>
</div>
<div id="pre-requisites" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h3>
<ul>
<li><p>Basic knowledge of R syntax</p></li>
<li><p>Familiarity with the SummarizedExperiment class</p></li>
<li><p>Familiarity with the GenomicRanges class</p></li>
<li><p>Familiarity with high-throughput genomic assays such as microarrays and next-generation sequencing</p></li>
<li><p>Familiarity with the biological definition of single nucleotide polymorphism (SNP) and copy number variation (CNV)</p></li>
</ul>
</div>
<div id="workshop-participation" class="section level3">
<h3 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h3>
<p>Execution of example code and hands-on practice</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<ul>
<li><a href="http://bioconductor.org/packages/RaggedExperiment">RaggedExperiment</a></li>
<li><a href="http://bioconductor.org/packages/CNVRanger">CNVRanger</a></li>
<li><a href="http://bioconductor.org/packages/regioneR">regioneR</a></li>
<li><a href="http://bioconductor.org/packages/PureCN">PureCN</a></li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Data representation and manipulation</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Integrative downstream analysis (eQTL, GWAS, …)</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Allele-specific CN analysis in cancer</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning Goals</h3>
<ul>
<li>get familiar with elementary concepts of CNV analysis</li>
<li>learn how to efficiently represent, access, and manipulate CNV data in Bioconductor data structures</li>
<li>get familiar with different strategies for summarizing individual CNV calls across a population</li>
<li>learn how to assess the significance of overlaps between CNVs and functional genomic regions</li>
<li>learn how carry out association analysis with gene expression and quantitative phenotypes</li>
<li>get familiar with allele-specific absolute CN analysis of genomic cancer data</li>
</ul>
</div>
<div id="specific-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#specific-objectives" class="anchor"></a>Specific objectives</h3>
<ul>
<li>understand how CNVs can be experimentally detected and computationally inferred from SNP arrays and next-generation sequencing data</li>
<li>learn how to use <code>GRangesList</code> and <code>RaggedExperiment</code> to represent, access, and manipulate CNV data</li>
<li>understand different strategies for finding recurrent CNV regions in a population, including density trimming, reciprocal overlap, and recurrence significance estimation</li>
<li>learn how to use the <a href="http://bioconductor.org/packages/regioneR">regioneR</a> package to assess the significance of overlaps between CNVs and functional genomic regions such as genes, promoters, and enhancers.</li>
<li>learn how to carry out eQTL analysis for CNV and RNA-seq data</li>
<li>learn how to carry out a GWAS analysis for CNV and quantitative phenotype data</li>
<li>learn how to estimate tumor purity and ploidy from absolute CN analysis with <a href="http://bioconductor.org/packages/PureCN">PureCN</a>
</li>
</ul>
</div>
</div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Copy number variation (CNV) is a frequently observed deviation from the diploid state due to duplication or deletion of genomic regions. CNVs can be experimentally detected based on comparative genomic hybridization, and computationally inferred from SNP-arrays or next-generation sequencing data. These technologies for CNV detection have in common that they report, for each sample under study, genomic regions that are duplicated or deleted with respect to a reference. Such regions are denoted as <em>CNV calls</em> in the following and will be considered the starting point for analysis.</p>
</div>
<div id="representation-and-manipulation-of-cnv-data-with-raggedexperiment" class="section level2">
<h2 class="hasAnchor">
<a href="#representation-and-manipulation-of-cnv-data-with-raggedexperiment" class="anchor"></a>Representation and manipulation of CNV data with RaggedExperiment</h2>
<p><code>RaggedExperiment</code> is a flexible data representation for segmented copy number, somatic mutations such as represented in <code>.vcf</code> files, and other ragged array schema for genomic location data. Like the <code>GRangesList</code> class from <code>GenomicRanges</code>, <code>RaggedExperiment</code> can be used to represent <em>differing</em> genomic ranges on each of a set of samples. In fact, <code>RaggedExperiment</code> contains a <code>GRangesList</code>:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu">showClass</span>(<span class="st">"RaggedExperiment"</span>)</pre></body></html></div>
<pre><code>## Class "RaggedExperiment" [package "RaggedExperiment"]
## 
## Slots:
##                                                       
## Name:       assays      rowidx      colidx    metadata
## Class: GRangesList     integer     integer        list
## 
## Extends: "Annotated"</code></pre>
<div id="constructing-a-raggedexperiment-object" class="section level3">
<h3 class="hasAnchor">
<a href="#constructing-a-raggedexperiment-object" class="anchor"></a>Constructing a <code>RaggedExperiment</code> object</h3>
<p>We start with a toy example of two <code>GRanges</code> objects, providing ranges on two chromosomes in two samples:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">sample1</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">A</span> <span class="kw">=</span> <span class="st">"chr1:1-10:-"</span>, <span class="kw">B</span> <span class="kw">=</span> <span class="st">"chr1:8-14:+"</span>, <span class="kw">C</span> <span class="kw">=</span> <span class="st">"chr1:15-18:+"</span>),
    <span class="kw">score</span> <span class="kw">=</span> <span class="fl">3</span>:<span class="fl">5</span>, <span class="kw">type</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"germline"</span>, <span class="st">"somatic"</span>, <span class="st">"germline"</span>))
<span class="no">sample2</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">D</span> <span class="kw">=</span> <span class="st">"chr1:1-10:-"</span>, <span class="kw">E</span> <span class="kw">=</span> <span class="st">"chr1:11-18:+"</span>),
    <span class="kw">score</span> <span class="kw">=</span> <span class="fl">11</span>:<span class="fl">12</span>, <span class="kw">type</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"germline"</span>, <span class="st">"somatic"</span>))</pre></body></html></div>
<p>Include column data <code>colData</code> to describe the samples:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">colDat</span> <span class="kw">&lt;-</span> <span class="fu">DataFrame</span>(<span class="kw">id</span><span class="kw">=</span><span class="fl">1</span>:<span class="fl">2</span>, <span class="kw">status</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"control"</span>, <span class="st">"case"</span>)))</pre></body></html></div>
<p>The <code>RaggedExperiment</code> can be constructed from individual <code>Granges</code>:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r">(<span class="no">ragexp</span> <span class="kw">&lt;-</span> <span class="fu">RaggedExperiment</span>(
    <span class="kw">sample1</span> <span class="kw">=</span> <span class="no">sample1</span>,
    <span class="kw">sample2</span> <span class="kw">=</span> <span class="no">sample2</span>,
    <span class="kw">colData</span> <span class="kw">=</span> <span class="no">colDat</span>))</pre></body></html></div>
<pre><code>## class: RaggedExperiment 
## dim: 5 2 
## assays(2): score type
## rownames(5): A B C D E
## colnames(2): sample1 sample2
## colData names(2): id status</code></pre>
<p>Or from a <code>GRangesList</code>:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">grl</span> <span class="kw">&lt;-</span> <span class="fu">GRangesList</span>(<span class="kw">sample1</span><span class="kw">=</span><span class="no">sample1</span>, <span class="kw">sample2</span><span class="kw">=</span><span class="no">sample2</span>)
<span class="no">ragexp2</span> <span class="kw">&lt;-</span> <span class="fu">RaggedExperiment</span>(<span class="no">grl</span>, <span class="kw">colData</span> <span class="kw">=</span> <span class="no">colDat</span>)
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="no">ragexp</span>, <span class="no">ragexp2</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>Note that the original ranges are is represented as the <code>rowRanges</code> of the <code>RaggedExperiment</code>:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu">rowRanges</span>(<span class="no">ragexp</span>)</pre></body></html></div>
<pre><code>## GRanges object with 5 ranges and 0 metadata columns:
##     seqnames    ranges strand
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   A     chr1      1-10      -
##   B     chr1      8-14      +
##   C     chr1     15-18      +
##   D     chr1      1-10      -
##   E     chr1     11-18      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="assay-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#assay-functions" class="anchor"></a><code>*Assay</code> functions</h3>
<p><code>RaggedExperiment</code> provides a flexible set of _*Assay_ methods to support transformation of data to matrix format with varying row dimensions.</p>
<div class="figure">
<img src="RaggedExperiment.png" alt="RaggedExperiment object schematic. Rows and columns represent genomic ranges and samples, respectively. Assay operations can be performed with (from left to right) compactAssay, qreduceAssay, and sparseAssay." width="\maxwidth"><p class="caption">
RaggedExperiment object schematic. Rows and columns represent genomic ranges and samples, respectively. Assay operations can be performed with (from left to right) compactAssay, qreduceAssay, and sparseAssay.
</p>
</div>
<p>The four main <em>Assay</em> functions for converting to matrix are:</p>
<ul>
<li>
<a href="#sparseassay">sparseAssay</a>: leave ranges exactly as-is</li>
<li>
<a href="#compactassay">compactAssay</a>: combine identical ranges</li>
<li>
<a href="#disjoinassay">disjoinAssay</a>: disjoin ranges that overlap across samples</li>
<li>
<a href="#qreduceassay">qreduceAssay</a>: find overlaps with provided “query” ranges</li>
</ul>
<p>These each have a corresponding function for conversion to <a href="#Conversion%20to%20RangedSummarizedExperiment">RangedSummarizedExperiment</a>.</p>
<div id="sparseassay" class="section level4">
<h4 class="hasAnchor">
<a href="#sparseassay" class="anchor"></a>sparseAssay</h4>
<p>The most straightforward matrix representation of a <code>RaggedExperiment</code> will return a matrix with the number of rows equal to the total number of ranges defined across all samples. <em>i.e.</em> the 5 rows of the <code>sparseAssay</code> result:</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu">sparseAssay</span>(<span class="no">ragexp</span>)</pre></body></html></div>
<pre><code>##   sample1 sample2
## A       3      NA
## B       4      NA
## C       5      NA
## D      NA      11
## E      NA      12</code></pre>
<p>correspond to the ranges of the unlisted <code>GRangesList</code>:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">grl</span>)</pre></body></html></div>
<pre><code>## GRanges object with 5 ranges and 2 metadata columns:
##             seqnames    ranges strand |     score        type
##                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   sample1.A     chr1      1-10      - |         3    germline
##   sample1.B     chr1      8-14      + |         4     somatic
##   sample1.C     chr1     15-18      + |         5    germline
##   sample2.D     chr1      1-10      - |        11    germline
##   sample2.E     chr1     11-18      + |        12     somatic
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>The rownames of the <code>sparseAssay</code> result are equal to the names of the <code>GRanges</code> elements. The values in the matrix returned by <code>sparseAssay</code> correspond to the first columns of the <code>mcols</code> of each <code>GRangesList</code> element, in this case the “score” column.</p>
<p>Note, this is the default <code>assay()</code> method of <code>RaggedExperiment</code>:</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu">assay</span>(<span class="no">ragexp</span>, <span class="st">"score"</span>)</pre></body></html></div>
<pre><code>##   sample1 sample2
## A       3      NA
## B       4      NA
## C       5      NA
## D      NA      11
## E      NA      12</code></pre>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu">assay</span>(<span class="no">ragexp</span>, <span class="st">"type"</span>)</pre></body></html></div>
<pre><code>##   sample1    sample2   
## A "germline" NA        
## B "somatic"  NA        
## C "germline" NA        
## D NA         "germline"
## E NA         "somatic"</code></pre>
</div>
<div id="compactassay" class="section level4">
<h4 class="hasAnchor">
<a href="#compactassay" class="anchor"></a>compactAssay</h4>
<p>The dimensions of the <code>compactAssay</code> result differ from that of the <code>sparseAssay</code> result only if there are identical ranges in different samples. Identical ranges are placed in the same row in the output. Ranges with any difference in start, end, or strand, will be kept on different rows. Non-disjoint ranges are <strong>not</strong> collapsed.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu">compactAssay</span>(<span class="no">ragexp</span>)</pre></body></html></div>
<pre><code>##              sample1 sample2
## chr1:8-14:+        4      NA
## chr1:11-18:+      NA      12
## chr1:15-18:+       5      NA
## chr1:1-10:-        3      11</code></pre>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu">compactAssay</span>(<span class="no">ragexp</span>, <span class="st">"type"</span>)</pre></body></html></div>
<pre><code>##              sample1    sample2   
## chr1:8-14:+  "somatic"  NA        
## chr1:11-18:+ NA         "somatic" 
## chr1:15-18:+ "germline" NA        
## chr1:1-10:-  "germline" "germline"</code></pre>
<p>Note that row names are constructed from the ranges, and the names of the <code>GRanges</code> vectors are lost, unlike in the <code>sparseAssay</code> result.</p>
</div>
<div id="disjoinassay" class="section level4">
<h4 class="hasAnchor">
<a href="#disjoinassay" class="anchor"></a>disjoinAssay</h4>
<p>This function is similar to <code>compactAssay</code> except the rows are <em>disjoint</em><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> ranges. Elements of the matrix are summarized by applying the <code>simplifyDisjoin</code> functional argument to assay values of overlapping ranges.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu">disjoinAssay</span>(<span class="no">ragexp</span>, <span class="kw">simplifyDisjoin</span> <span class="kw">=</span> <span class="no">mean</span>)</pre></body></html></div>
<pre><code>##              sample1 sample2
## chr1:8-10:+        4      NA
## chr1:11-14:+       4      12
## chr1:15-18:+       5      12
## chr1:1-10:-        3      11</code></pre>
</div>
<div id="qreduceassay" class="section level4">
<h4 class="hasAnchor">
<a href="#qreduceassay" class="anchor"></a>qreduceAssay</h4>
<p>The <code>qreduceAssay</code> function is the most complicated but likely the most useful of the <code>RaggedExperiment</code> <em>Assay</em> functions. It requires you to provide a <code>query</code> argument that is a <code>GRanges</code> vector, and the rows of the resulting matrix correspond to the elements of this <code>GRanges</code>. The returned matrix will have dimensions <code><a href="https://rdrr.io/r/base/length.html">length(query)</a></code> by <code><a href="https://rdrr.io/r/base/nrow.html">ncol(x)</a></code>. Elements of the resulting matrix correspond to the overlap of the <em>i</em> th <code>query</code> range in the <em>j</em> th sample, summarized according to the <code>simplifyReduce</code> functional argument. This can be useful, for example, to calculate per-gene copy number or mutation status by providing the genomic ranges of every gene as the <code>query</code>.</p>
<p>The <code>simplifyReduce</code> argument in <code>qreduceAssay</code> allows the user to summarize overlapping regions with a custom method for the given “query” region of interest. We provide one for calculating a weighted average score per query range, where the weight is proportional to the overlap widths between overlapping ranges and a query range.</p>
<p><em>Note</em> that there are three arguments to this function. See the documentation for additional details.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">weightedmean</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">scores</span>, <span class="no">ranges</span>, <span class="no">qranges</span>)
{
    <span class="no">isects</span> <span class="kw">&lt;-</span> <span class="fu">pintersect</span>(<span class="no">ranges</span>, <span class="no">qranges</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">scores</span> * <span class="fu">width</span>(<span class="no">isects</span>)) / <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu">width</span>(<span class="no">isects</span>))
}</pre></body></html></div>
<p>The call to <code>qreduceAssay</code> calculates the overlaps between the ranges of each sample:</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">grl</span></pre></body></html></div>
<pre><code>## GRangesList object of length 2:
## $sample1
## GRanges object with 3 ranges and 2 metadata columns:
##     seqnames    ranges strand |     score        type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   A     chr1      1-10      - |         3    germline
##   B     chr1      8-14      + |         4     somatic
##   C     chr1     15-18      + |         5    germline
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $sample2
## GRanges object with 2 ranges and 2 metadata columns:
##     seqnames    ranges strand |     score        type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   D     chr1      1-10      - |        11    germline
##   E     chr1     11-18      + |        12     somatic
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>with the query ranges (an arbitrary set is defined here for demonstration): First create a demonstration “query” region of interest:</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r">(<span class="no">query</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr1:1-14:-"</span>, <span class="st">"chr1:15-18:+"</span>)))</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     chr1      1-14      -
##   [2]     chr1     15-18      +
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>using the <code>simplifyReduce</code> function to resolve overlapping ranges and return a matrix with rows corresponding to the query:</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu">qreduceAssay</span>(<span class="no">ragexp</span>, <span class="no">query</span>, <span class="kw">simplifyReduce</span> <span class="kw">=</span> <span class="no">weightedmean</span>)</pre></body></html></div>
<pre><code>##              sample1 sample2
## chr1:1-14:-        3      11
## chr1:15-18:+       5      12</code></pre>
</div>
</div>
<div id="conversion-to-rangedsummarizedexperiment" class="section level3">
<h3 class="hasAnchor">
<a href="#conversion-to-rangedsummarizedexperiment" class="anchor"></a>Conversion to RangedSummarizedExperiment</h3>
<p>These methods all have corresponding methods to return a <code>RangedSummarizedExperiment</code> and preserve the <code>colData</code>:</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu">sparseSummarizedExperiment</span>(<span class="no">ragexp</span>)
<span class="fu">compactSummarizedExperiment</span>(<span class="no">ragexp</span>)
<span class="fu">disjoinSummarizedExperiment</span>(<span class="no">ragexp</span>, <span class="kw">simplify</span> <span class="kw">=</span> <span class="no">mean</span>)
<span class="fu">qreduceSummarizedExperiment</span>(<span class="no">ragexp</span>, <span class="kw">query</span><span class="kw">=</span><span class="no">query</span>, <span class="kw">simplify</span><span class="kw">=</span><span class="no">weightedmean</span>)</pre></body></html></div>
<p>Please see the <code>RaggedExperiment</code> vignette for <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/RaggedExperiment/inst/doc/RaggedExperiment.html">more details</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>A <em>disjoint</em> set of ranges has no overlap between any ranges of the set.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
